llext: make unresolved symbol errors fatal

lyakh opened the pull request:
With experience it becomes clear, that failing to resolve symbols during the linking process is likely fatal for the module loading and a simple warning isn't enough. Fail loading instead.

lyakh committed:
llext: make unresolved symbol errors fatal

With experience it becomes clear, that failing to resolve symbols
during the linking process is likely fatal for the module loading and
a simple warning isn't enough. Fail loading instead.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

pillo79 reviewed (commented):
Thanks! LGTM, two considerations:
1) this will stop at the first issue. Would you consider stopping at the end of linking, so _all_ unresolved symbols will be printed at once?
2) can we keep the non-PLT case in sync? I think that already fails early and if you agree with 1) I'd prefer to do the same change there as well.

lyakh committed:
llext: move 2 prototypes to llext_internal.h

arch_elf_relocate_global() and arch_elf_relocate_local() aren't LLEXT
API functions, they're internal implementation functions, move them
to llext_internal.h

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

pillo79 reviewed (commented):
Thanks for the changes, they mostly LGTM. However, your patch highlighted an unintentional but still terrible situation, where some control flow choices depend on debug being active! :scream:

The following changes should fix the issue while keeping the spirit of the PR. What do you think?

lyakh commented on subsys/llext/llext_link.c:
```diff
@@ -512,7 +532,10 @@ int llext_link(struct llext_loader *ldr, struct llext *ext, const struct llext_l
 				if (ret != 0) {
 					LOG_ERR("Could not find symbol %s!", name);
-					return ret;
+					if (!link_err) {
+						link_err = ret;
+					}
+					continue;
```
the `return` in 504 / 524 should be modified too then

pillo79 reviewed (approved):
Thanks, +1 to moving the arch specific functions in `llext_internal.h`!
~If any change is requested, you should also move the `arch_elf_relocate` function along.~ EDIT: nevermind, did not see there was already an approval :slightly_smiling_face:
